### PR TITLE
added mesos authentication

### PR DESCRIPTION
**Pull request by sebbrandt87**
### Required for all PRs:
- [ ] CHANGELOG.md updated (we recommend not updating this until the PR has been approved by a maintainer)
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)
- [x] README.md updated (if adding a new plugin)

### CODE REVIEW

**Comment by sparrc** [comment]
what is this debug.test file?

File: internal/config/testdata/telegraf-agent.toml [hunk 1]
```diff
@@ -190,6 +190,12 @@
   timeout = 100
   # A list of Mesos masters, default value is localhost:5050.
   masters = ["localhost:5050"]
+  # Default protocol for communication. Use https if you have authentication!
+  protocol = "http"
```
Comment (sparrc) [inline internal/config/testdata/telegraf-agent.toml]: I would prefer if the protocol scheme were prepended to the server, like `masters = ["http://localhost:5050"]`. You can assume `http` if there is no protocol

Comment (sebbrandt87) [inline internal/config/testdata/telegraf-agent.toml]: For sure I can change it to that. I just intended that if there is an active TLS it is active for a hole cluster and not just defined per host.

**Comment by sebbrandt87** [comment]
anything more to do?

File: plugins/inputs/mesos/mesos.go [hunk 2]
```diff
@@ -91,9 +103,16 @@ func (m *Mesos) SetDefaults() {
 		log.Println("I! [mesos] Missing timeout value, setting default value (100ms)")
 		m.Timeout = 100
 	}
+
+	if len(m.Username) == 0 {
```
Comment (sparrc) [inline plugins/inputs/mesos/mesos.go]: these if-statements are unnecessary

File: plugins/inputs/mesos/mesos.go [hunk 3]
```diff
@@ -42,7 +49,11 @@ var sampleConfig = `
   ## Timeout, in ms.
   timeout = 100
   ## A list of Mesos masters.
-  masters = ["localhost:5050"]
+  masters = ["http://localhost:5050"]
+	# Authentication username
```
Comment (sparrc) [inline plugins/inputs/mesos/mesos.go]: these tabs should be two spaces instead

File: plugins/inputs/mesos/mesos.go [hunk 4]
```diff
@@ -418,29 +437,63 @@ var client = &http.Client{
...
 
-	resp, err := client.Get("http://" + address + "/monitor/statistics?timeout=" + ts)
+	req, err := http.NewRequest(
+		"GET",
+		endpoint["scheme"]+"://"+endpoint["host"]+":"+endpoint["port"]+"/monitor/statistics?timeout="+ts,
+		nil,
+	)
+
+	if len(username) > 0 && len(password) > 0 {
+		req.Header.Add("Authorization", "Basic "+basicAuth(username, password))
```
Comment (sparrc) [inline plugins/inputs/mesos/mesos.go]: there is a Request.SetBasicAuth function that I think you should be using instead: https://golang.org/pkg/net/http/#Request.SetBasicAuth

File: plugins/inputs/mesos/mesos.go [hunk 5]
```diff
@@ -418,29 +430,62 @@ var client = &http.Client{
 	Timeout:   time.Duration(4 * time.Second),
 }
 
+func basicAuth(username, password string) string {
+	auth := username + ":" + password
+	return base64.URLEncoding.EncodeToString([]byte(auth))
+}
+
+func getEndpointParams(address string) map[string]string {
```
Comment (sparrc) [inline plugins/inputs/mesos/mesos.go]: This function seems a bit unecessary. Basically all you are doing is taking apart the given URL into three different parts, and then you later just put it back together exactly as it was already provided.

File: plugins/inputs/mesos/mesos.go [hunk 6]
```diff
@@ -418,29 +430,62 @@ var client = &http.Client{
...
-		"server": host,
+		"server": endpoint["host"],
 	}
 
 	ts := strconv.Itoa(m.Timeout) + "ms"
 
-	resp, err := client.Get("http://" + address + "/monitor/statistics?timeout=" + ts)
+	req, err := http.NewRequest(
+		"GET",
+		endpoint["scheme"]+"://"+endpoint["host"]+":"+endpoint["port"]+"/monitor/statistics?timeout="+ts,
```
Comment (sparrc) [inline plugins/inputs/mesos/mesos.go]: all you need to do with the url is prepend `http://` if it's not already there, use `strings.HasPrefix`

File: plugins/inputs/mesos/mesos.go [hunk 7]
```diff
@@ -418,29 +430,62 @@ var client = &http.Client{
 	Timeout:   time.Duration(4 * time.Second),
 }
 
+func basicAuth(username, password string) string {
```
Comment (sparrc) [inline plugins/inputs/mesos/mesos.go]: delete this function as it's not being used anymore

File: plugins/inputs/mesos/mesos.go (hunk 6 above)
Comment (sebbrandt87) [inline plugins/inputs/mesos/mesos.go]: I already do within getEndpointParams

**Comment by sparrc** [comment]
I'd prefer if you parsed the URL at a higher level, in the Gather function. You can set an object on the Mesos struct, and then only set it in Gather if it is non-nil.

Also, just use Go's builtin net.URL type rather than creating your own. You will need to prepend `http://` before calling `net.Parse()`. You can use the net.URL fields for UserInfo & Path to create the object that you need, and then later use the String() function.

**Comment by sebbrandt87** [comment]
so what I am going to remove is for sure the concating of the string and do it with sprintf.
I implemented url.Parse() into the getEndpointParams as first operation if non http is present at all.
And I will have a look again into the url fields. Didn't recognized that before sry.

**Comment by sebbrandt87** [comment]
is it fitting now?

File: plugins/inputs/mesos/mesos.go [hunk 8]
```diff
@@ -425,22 +455,22 @@ type TaskStats struct {
...
-		address = address + defaultPort
-	}
+	address, host := prepareAddress(address)
 
 	tags := map[string]string{
 		"server": host,
 	}
 
-	ts := strconv.Itoa(m.Timeout) + "ms"
+	requestURL := fmt.Sprintf("%s/monitor/statistics?timeout=%sms", address, strconv.Itoa(m.Timeout))
```
Comment (sparrc) [inline plugins/inputs/mesos/mesos.go]: `%sms` could be changed to `%dms` and you can get rid of the strconv call

File: plugins/inputs/mesos/mesos.go [hunk 9]
```diff
@@ -42,7 +49,11 @@ var sampleConfig = `
   ## Timeout, in ms.
   timeout = 100
   ## A list of Mesos masters.
-  masters = ["localhost:5050"]
+  masters = ["http://localhost:5050"]
+  # Authentication username
```
Comment (sparrc) [inline plugins/inputs/mesos/mesos.go]: indicate that these use http basic auth

File: plugins/inputs/mesos/mesos.go [hunk 10]
```diff
@@ -93,7 +105,25 @@ func (m *Mesos) SetDefaults() {
...
 
-// Gather() metrics from given list of Mesos Masters
+func prepareAddress(address string) (prepared, host string) {
+	if strings.HasPrefix(address, "http") != true {
+		address = "http://" + address
+	}
+
+	u, err := url.Parse(address)
+	if err != nil {
+		panic(err)
```
Comment (sparrc) [inline plugins/inputs/mesos/mesos.go]: do not panic!! you need to handle the errors here gracefully

File: plugins/inputs/mesos/mesos.go [hunk 11]
```diff
@@ -93,7 +105,25 @@ func (m *Mesos) SetDefaults() {
...
+	}
+
+	host, _, err = net.SplitHostPort(u.Host)
+	if err != nil {
+		panic(err)
```
Comment (sparrc) [inline plugins/inputs/mesos/mesos.go]: handle panic gracefully

File: plugins/inputs/mesos/mesos.go [hunk 12]
```diff
@@ -93,7 +105,25 @@ func (m *Mesos) SetDefaults() {
 	}
 }
 
-// Gather() metrics from given list of Mesos Masters
+func prepareAddress(address string) (prepared, host string) {
```
Comment (sparrc) [inline plugins/inputs/mesos/mesos.go]: this function has an odd signature, I think it should return `(*url.URL, error)`

File: plugins/inputs/mesos/mesos.go [hunk 13]
```diff
@@ -93,7 +105,25 @@ func (m *Mesos) SetDefaults() {
...
+	if strings.HasPrefix(address, "http") != true {
+		address = "http://" + address
+	}
+
+	u, err := url.Parse(address)
+	if err != nil {
+		panic(err)
+	}
+
+	host, _, err = net.SplitHostPort(u.Host)
```
Comment (sparrc) [inline plugins/inputs/mesos/mesos.go]: this seems unnecessary, why not include the port in the tag?

File: plugins/inputs/mesos/mesos.go (hunk 8 above)
Comment (sparrc) [inline plugins/inputs/mesos/mesos.go]: if you return *url.URL from prepareAddress, you can construct this full request URI using Go builtin URL parsing. See [url.Values](https://golang.org/pkg/net/url/#Values) and [url.URL.Path](https://golang.org/pkg/net/url/#URL)

**Comment by sparrc** [comment]
thanks for the contribution and apologies for the initial delay in reviewing, feel free to re-open if you can find time to make the requested changes.

**Pull request closed without merging**